xmppserver/src/main/java/org/jivesoftware/openfire/muc/spi/LocalMUCUser.java

```diff
@@ -574,6 +575,76 @@ private void processDecliningInvitationMessage(
                                          info.elementTextTrim("reason"), packet.getFrom());
     }
 
+    /**
+     * This method will parse a packet and checks if it is a MEMBER! list retrieval packet or not. 
+     * @param packet IQ
+     * @return true, if it is or false if not
+     */
+    private boolean isListRetrieval(IQ packet)
+    {
+        Element query = packet.getChildElement();
+        if (query!=null&&query.getName().equalsIgnoreCase("query"))
+        {
+            if (query.getNamespace()!=null&&query.getNamespace().getURI().equals("http://jabber.org/protocol/muc#admin"))
+            {
+                Element item = query.element("item");
+                if (item!=null&&item.attributeValue("affiliation")!=null&&
+                    item.attributeValue("affiliation").equalsIgnoreCase("member"))
+                {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Checks if packet sender (non-occupant) is allowed to retrieve a memberlist.
+     * Sender must have a member privileges at minimum to be allowed. 
+     * @param packet IQ
+     * @return true, if sender is allowed to retrieve or false if not
+     */
+    private boolean hasPermissionToRetrieveList(IQ packet)
+    {
+        JID from = packet.getFrom();
+        MUCRoom room = server.hasChatRoom(packet.getTo().getNode())?server.getChatRoom(packet.getTo().getNode()):null;
+        return  room!=null&&(room.getOwners().contains(from.asBareJID())||
```

Comment:
I believe that there is room for an additional feature. Although the XEP specifies that members SHOULD be able to retrieve the memberlist of a members-only room, it does not explicitly say what to do for rooms that are not members-only. Depending on the configuration of the room, all members are exposed through service discovery. I think that it is safe to say that when a sender is allowed to retrieve the members of a room through service discovery, then it should also be allowed to perform the memberlist query. The relevant code is in `org.jivesoftware.openfire.muc.spi.MultiUserChatServiceImpl#getItems(java.lang.String, java.lang.String, org.xmpp.packet.JID)`

Reply:
I would check if it is a memberonly room then check like above and on the other hand everybody is allows to retrieve the list... that is more simple...